### Data Understanding

In [1]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [9]:
# Load the dataset
train = pd.read_csv("../data/raw/Train.csv")
test = pd.read_csv("../data/raw/Test.csv")

#### 1. Preview the data

In [10]:
train.head()

,ID,label,VH_01,VV_01,blue_01,green_01,nir_01,nira_01,re1_01,re2_01,...,blue_12,green_12,nir_12,nira_12,re1_12,re2_12,re3_12,red_12,swir1_12,swir2_12
0,ID_TR_NEW_XVGKFMLNRJ,0,-29.099645,-22.471573,1665,1719,1367,1270,1689,1416,...,1639,1826,1395,1384,1736,1449,1436,1626,1307,1216
1,ID_TR_NEW_GP8KNSWVP6,0,-19.470574,-10.752340,1579,1740,2245,2231,2060,2147,...,1490,1622,2090,2109,1924,1941,2020,1779,2298,2131
2,ID_TR_NEW_87X3957MVS,1,-20.964854,-8.792675,1850,2345,3664,3470,2935,3384,...,1438,1673,1335,1289,1588,1377,1307,1503,1237,1168
3,ID_TR_NEW_T4JMRPKHS3,0,-18.728208,-4.689202,1279,1297,1713,1698,1527,1531,...,1405,1550,2454,2397,1894,2040,2179,1748,2574,1977
4,ID_TR_NEW_2CTUQQ8KLU,0,-29.630941,-23.370157,1625,1741,1339,1258,1658,1362,...,1582,1856,1365,1306,1627,1370,1367,1641,1219,1147


#### 2. Number of rows and columns

In [14]:
print()

rows = train.shape[0]
columns = train.shape[1]

print("Train Dataset")
print()

print(f"Rows: {rows}")
print(f"Columns: {columns}")


Train Dataset

Rows: 1821
Columns: 146


In [15]:
print()

rows = test.shape[0]
columns = test.shape[1]

print("Test Dataset")
print()
print(f"Rows: {rows}")
print(f"Columns: {columns}")


Test Dataset

Rows: 1030
Columns: 145


This indicates the training dataset includes one additional column, which is likely the target variable.

#### 3. Column names

In [17]:
train.columns.tolist()

['ID',
 'label',
 'VH_01',
 'VV_01',
 'blue_01',
 'green_01',
 'nir_01',
 'nira_01',
 're1_01',
 're2_01',
 're3_01',
 'red_01',
 'swir1_01',
 'swir2_01',
 'VH_02',
 'VV_02',
 'blue_02',
 'green_02',
 'nir_02',
 'nira_02',
 're1_02',
 're2_02',
 're3_02',
 'red_02',
 'swir1_02',
 'swir2_02',
 'VH_03',
 'VV_03',
 'blue_03',
 'green_03',
 'nir_03',
 'nira_03',
 're1_03',
 're2_03',
 're3_03',
 'red_03',
 'swir1_03',
 'swir2_03',
 'VH_04',
 'VV_04',
 'blue_04',
 'green_04',
 'nir_04',
 'nira_04',
 're1_04',
 're2_04',
 're3_04',
 'red_04',
 'swir1_04',
 'swir2_04',
 'VH_05',
 'VV_05',
 'blue_05',
 'green_05',
 'nir_05',
 'nira_05',
 're1_05',
 're2_05',
 're3_05',
 'red_05',
 'swir1_05',
 'swir2_05',
 'VH_06',
 'VV_06',
 'blue_06',
 'green_06',
 'nir_06',
 'nira_06',
 're1_06',
 're2_06',
 're3_06',
 'red_06',
 'swir1_06',
 'swir2_06',
 'VH_07',
 'VV_07',
 'blue_07',
 'green_07',
 'nir_07',
 'nira_07',
 're1_07',
 're2_07',
 're3_07',
 'red_07',
 'swir1_07',
 'swir2_07',
 'VH_08',
 'VV_08

In [18]:
print()
test.columns.tolist()

['ID',
 'VH_01',
 'VV_01',
 'blue_01',
 'green_01',
 'nir_01',
 'nira_01',
 're1_01',
 're2_01',
 're3_01',
 'red_01',
 'swir1_01',
 'swir2_01',
 'VH_02',
 'VV_02',
 'blue_02',
 'green_02',
 'nir_02',
 'nira_02',
 're1_02',
 're2_02',
 're3_02',
 'red_02',
 'swir1_02',
 'swir2_02',
 'VH_03',
 'VV_03',
 'blue_03',
 'green_03',
 'nir_03',
 'nira_03',
 're1_03',
 're2_03',
 're3_03',
 'red_03',
 'swir1_03',
 'swir2_03',
 'VH_04',
 'VV_04',
 'blue_04',
 'green_04',
 'nir_04',
 'nira_04',
 're1_04',
 're2_04',
 're3_04',
 'red_04',
 'swir1_04',
 'swir2_04',
 'VH_05',
 'VV_05',
 'blue_05',
 'green_05',
 'nir_05',
 'nira_05',
 're1_05',
 're2_05',
 're3_05',
 'red_05',
 'swir1_05',
 'swir2_05',
 'VH_06',
 'VV_06',
 'blue_06',
 'green_06',
 'nir_06',
 'nira_06',
 're1_06',
 're2_06',
 're3_06',
 'red_06',
 'swir1_06',
 'swir2_06',
 'VH_07',
 'VV_07',
 'blue_07',
 'green_07',
 'nir_07',
 'nira_07',
 're1_07',
 're2_07',
 're3_07',
 'red_07',
 'swir1_07',
 'swir2_07',
 'VH_08',
 'VV_08',
 'blue_

The label column is present only in the training dataset, indicating it is the target variable that the model will learn to predict.

#### 4. Dataset Information

In [24]:
train.info()
print()

<class 'pandas.DataFrame'>
RangeIndex: 1821 entries, 0 to 1820
Columns: 146 entries, ID to swir2_12
dtypes: float64(24), int64(121), str(1)
memory usage: 2.0 MB



In [23]:
test.info()

<class 'pandas.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Columns: 145 entries, ID to swir2_12
dtypes: float64(24), int64(120), str(1)
memory usage: 1.1 MB


#### 5. Statistical Overview

In [29]:
print()
train.describe().T

,count,mean,std,min,25%,50%,75%,max
label,1821.0,0.403624,0.490759,0.000000,0.000000,0.000000,1.000000,1.000000
VH_01,1821.0,-25.379836,5.586461,-48.143182,-29.510694,-23.945252,-21.090626,-10.404720
VV_01,1821.0,-15.136799,5.703961,-31.912804,-20.761611,-13.290349,-10.872466,6.227268
blue_01,1821.0,1651.551345,274.479438,1070.000000,1454.000000,1627.000000,1814.000000,3070.000000
green_01,1821.0,1845.806150,353.859045,1094.000000,1613.000000,1815.000000,2081.000000,3788.000000
...,...,...,...,...,...,...,...,...
re2_12,1821.0,2175.002746,808.934218,1034.000000,1530.000000,2106.000000,2527.000000,6201.000000
re3_12,1821.0,2237.421746,895.571444,1064.000000,1525.000000,2166.000000,2613.000000,6945.000000
red_12,1821.0,1782.470071,317.300419,1061.000000,1571.000000,1777.000000,1969.000000,3869.000000
swir1_12,1821.0,2249.512356,854.570513,1048.000000,1218.000000,2509.000000,2938.000000,4505.000000


The numerical features have different ranges and levels of variation, so feature scaling may be needed before training the model. Some features also have large differences between the 75th percentile and maximum values, which may indicate the presence of outliers. These differences show the need for data preprocessing and further feature analysis to improve model performance.

#### 6. Outlier Detection

In [39]:
print()
# Select numerical features
num_cols = train.select_dtypes(include=['int64', 'float64'])

# Calculate IQR
Q1 = num_cols.quantile(0.25)
Q3 = num_cols.quantile(0.75)
IQR = Q3 - Q1

# Detect outliers
outliers = ((num_cols < (Q1 - 1.5 * IQR)) | 
            (num_cols > (Q3 + 1.5 * IQR)))

# Count outliers
outlier_count = outliers.sum()

# Display results
print("Outlier Count per Feature:")
print(outlier_count[outlier_count > 0])

print()


Outlier Count per Feature:
VH_01       10
VV_01        1
blue_01     30
green_01    15
nir_01      10
            ..
nira_12     57
re1_12       9
re2_12      69
re3_12      74
red_12      32
Length: 110, dtype: int64



In [42]:
print()
num_features = train.select_dtypes(include=['int64', 'float64']).shape[1]

print("Number of numerical features:", num_features)

print()


Number of numerical features: 145



Outliers were detected in 110 out of 145 numerical features, showing that many features contain extreme values that should be reviewed during preprocessing.

#### 7. Missing Data Overview

In [34]:
train.isnull().sum()

ID          0
label       0
VH_01       0
VV_01       0
blue_01     0
           ..
re2_12      0
re3_12      0
red_12      0
swir1_12    0
swir2_12    0
Length: 146, dtype: int64

The dataset does not contain any missing values, meaning all features have complete information and no missing value handling is needed.

#### 8. Duplicate Values Check

In [35]:
train.duplicated().sum()

np.int64(0)

The dataset contains no duplicate records, ensuring that all observations are unique.

#### 9. Target Variable Distribution

In [43]:
train['label'].value_counts()

label
0    1086
1     735
Name: count, dtype: int64

In [44]:
print()
train['label'].value_counts(normalize=True) * 100

label
0    59.637562
1    40.362438
Name: proportion, dtype: float64

The dataset contains more samples of class 0 (59.6%) than class 1 (40.4%), showing a slight imbalance between the two classes.

#### 10. Numerical and Categorical Feature Identification

In [51]:
train.select_dtypes(include='number').columns

Index(['label', 'VH_01', 'VV_01', 'blue_01', 'green_01', 'nir_01', 'nira_01',
       're1_01', 're2_01', 're3_01',
       ...
       'blue_12', 'green_12', 'nir_12', 'nira_12', 're1_12', 're2_12',
       're3_12', 'red_12', 'swir1_12', 'swir2_12'],
      dtype='str', length=145)

In [47]:
import warnings
warnings.filterwarnings("ignore")
print()

train.select_dtypes(include='object').columns

Index(['ID'], dtype='str')

The dataset contains 145 numerical features and one string column (ID). Most features are numerical satellite measurements, so no categorical encoding is needed. The ID column only identifies each record and may not provide useful information for the model, so it can be considered for removal.

#### 11. Correlation Analysis

In [55]:
corr_matrix = train.select_dtypes(include='number').corr()

corr_matrix

,label,VH_01,VV_01,blue_01,green_01,nir_01,nira_01,re1_01,re2_01,re3_01,...,blue_12,green_12,nir_12,nira_12,re1_12,re2_12,re3_12,red_12,swir1_12,swir2_12
label,1.000000,-0.192640,-0.011321,0.485605,0.569295,-0.052713,-0.078865,0.322196,0.011128,-0.029308,...,0.402590,0.488047,-0.174110,-0.200675,0.185542,-0.119930,-0.151505,0.201003,-0.290883,-0.297883
VH_01,-0.192640,1.000000,0.842361,0.113222,0.073244,0.553460,0.577287,0.378826,0.522101,0.532633,...,-0.069148,-0.054316,0.425405,0.451924,0.284044,0.393167,0.403198,0.154267,0.556097,0.516476
VV_01,-0.011321,0.842361,1.000000,0.308253,0.284060,0.628502,0.646241,0.566992,0.614911,0.610812,...,0.047934,0.095768,0.435246,0.456306,0.392626,0.418539,0.418843,0.252700,0.544589,0.514167
blue_01,0.485605,0.113222,0.308253,1.000000,0.954550,0.417292,0.392799,0.829649,0.479025,0.429826,...,0.601229,0.638570,0.110334,0.092559,0.456789,0.153438,0.124330,0.518403,0.090148,0.112451
green_01,0.569295,0.073244,0.284060,0.954550,1.000000,0.474070,0.441802,0.867958,0.542186,0.492938,...,0.617677,0.709608,0.173428,0.150273,0.532620,0.223685,0.191715,0.529247,0.091468,0.093663
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
re2_12,-0.119930,0.393167,0.418539,0.153438,0.223685,0.620616,0.623019,0.427629,0.616559,0.625565,...,0.123684,0.303194,0.989849,0.987270,0.750948,1.000000,0.997409,0.246712,0.675020,0.519049
re3_12,-0.151505,0.403198,0.418843,0.124330,0.191715,0.617458,0.621228,0.404546,0.609541,0.620964,...,0.081993,0.259320,0.992861,0.992042,0.717652,0.997409,1.000000,0.210261,0.674489,0.516202
red_12,0.201003,0.154267,0.252700,0.518403,0.529247,0.292434,0.281276,0.515700,0.314548,0.292085,...,0.865781,0.806534,0.223661,0.214255,0.758282,0.246712,0.210261,1.000000,0.465770,0.529295
swir1_12,-0.290883,0.556097,0.544589,0.090148,0.091468,0.599823,0.623291,0.384964,0.561872,0.579006,...,0.137211,0.147832,0.709273,0.742157,0.640700,0.675020,0.674489,0.465770,1.000000,0.965559


The correlation analysis with the target variable shows that some satellite features have a stronger relationship with the label. Features such as green_01 and blue_01 show a positive relationship with the target, while features like swir1_12 and swir2_12 show a negative relationship. This indicates that different spectral bands may contribute differently to distinguishing between the two classes.